### PR TITLE
Create conversations

### PR DESCRIPTION
**Pull request by ivanfelfo**
Feature: Added conversations with messages

- User can create conversations with another user
- User can create messages in a conversation he is participating
- User can get all his conversations
- User can get all unread conversations with unread messages count

Missing: web sockets

### CODE REVIEW

File: app/models/conversation.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+class Conversation < ApplicationRecord
+  belongs_to :topic
+  has_many :messages, dependent: :destroy
+
+  validates :user_id1, presence: { on: :create, message: "can't be blank" }
+  validates :user_id2, presence: { on: :create, message: "can't be blank" }
```
Comment (martinjaimem) [inline app/models/conversation.rb:6]:
```suggestion
  validates :user_id1, presence: { on: :create }
  validates :user_id2, presence: { on: :create }
```

File: app/views/v1/conversations/index.json.jbuilder [hunk 2]
```diff
@@ -0,0 +1,3 @@
+json.conversations do
+  json.array! @conversations
```
Comment (martinjaimem) [inline app/views/v1/conversations/index.json.jbuilder:2]: you should call the `app/views/v1/conversations/create.json.jbuilder` here for each conversation

File: app/views/v1/conversations/show.json.jbuilder [hunk 3]
```diff
@@ -0,0 +1,3 @@
+json.conversation do
+  json.array! @conversation
```
Comment (martinjaimem) [inline app/views/v1/conversations/show.json.jbuilder:2]: you should call the `app/views/v1/conversations/create.json.jbuilder` here for each conversation

File: spec/models/conversation_spec.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  let(:new_conversation) { build(:conversation) }
+  subject { new_conversation.save! }
+
+  context 'when it\'s a valid instance' do
+    it 'saves the record' do
+      expect(subject).to eq(true)
+    end
+  end
+
+  context 'when is NOT a valid instance' do
```
Comment (martinjaimem) [inline spec/models/conversation_spec.rb:11]:
```suggestion
  context "when it's NOT a valid instance" do
```

File: spec/models/conversation_spec.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    context 'when trying to create a conversation with the same person' do
+      let(:new_conversation) { build(:conversation, user_id1: 1, user_id2: 1) }
+
+      it 'raises error' do
```
Comment (martinjaimem) [inline spec/models/conversation_spec.rb:15]:
```suggestion
      it 'raises an error' do
```

File: app/controllers/v1/conversations_controller.rb [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  class ConversationsController < ApiController
+    before_action :authenticate_v1_user!
+
+    def create
+      @conversation = Conversation.create!(params_conversation)
+    end
+
+    def index
+      @conversations = Conversation.where(user_id1: current_v1_user.id)
+                                   .or(Conversation.where(user_id2: current_v1_user.id))
```
Comment (juanmferreira93) [inline app/controllers/v1/conversations_controller.rb:11]: maybe you can move this to a private method

File: app/controllers/v1/conversations_controller.rb [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+      render json: { error: I18n.t('conversation.index.empty') }, status: :bad_request
+    end
+
+    def show
+      @conversation = conversation_array_show
+      @conversation.update!(read: true)
+      return unless @conversation.empty?
+
+      render json: { error: I18n.t('conversation.show.empty') }, status: :bad_request
+      elseif @conversation.messages.present?
```
Comment (juanmferreira93) [inline app/controllers/v1/conversations_controller.rb:23]: I do not know if it's for indentation or what, but I really do not understand this portion of the code, maybe avoid using `unless` with `elseif`

File: app/controllers/v1/conversations_controller.rb [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+      @conversation.messages.unread.each { |m| m.read = true }
+    end
+
+    def unread
+      @conversations = Conversation.where({ user_id1: current_v1_user.id,
+                                            read: false })
+                                   .or(Conversation
+                                              .where({ user_id2: current_v1_user.id, read: false }))
```
Comment (juanmferreira93) [inline app/controllers/v1/conversations_controller.rb:31]: this could be a private method, wdyt?

File: app/controllers/v1/conversations_controller.rb [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+      render json: { error: I18n.t('conversation.get_unread.empty') }, status: :bad_request
+    end
+
+    private
+
+    def conversation_array_show
+      Conversation.where({ id: params[:id],
+                           user_id1: current_v1_user.id })
+                  .or(Conversation.where({ id: params[:id],
+                                           user_id2: current_v1_user.id }))
```
Comment (juanmferreira93) [inline app/controllers/v1/conversations_controller.rb:43]: could you enhance the indentation for this method? it's a bit weird

File: app/channels/application_cable/connection.rb [hunk 10]
```diff
@@ -1,6 +1,19 @@
...
+    identified_by :v1_current_user
+
+    def connect
+      self.current_user = find_verified_user
+    end
+
+    private
+
+    def find_verified_user
+      if verified_user == User.find_by(id: cookies.encrypted[:user_id])
```
Comment (juanmferreira93) [inline app/channels/application_cable/connection.rb:12]: do we store user_id on cookies?

File: app/controllers/v1/messages_controller.rb [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    def check_user_message_conversation
+      unless conversation.user_id1 == current_v1_user.id ||
+             conversation.user_id2 == current_v1_user.id
+        render json: { error: I18n.t('message.index.empty') },
+               status: :bad_request
+      end
+    end
+
+    def conversation
+      Conversation.find(params[:id])
```
Comment (juanmferreira93) [inline app/controllers/v1/messages_controller.rb:27]: you could do something like: `@conversation ||= Conversation.find(params[:id])` so if you use it once, it avoid a new call to the DB

File: app/controllers/v1/registrations_controller.rb [hunk 12]
```diff
@@ -1,5 +1,3 @@
-# frozen_string_literal: true
-
```
Comment (juanmferreira93) [inline app/controllers/v1/registrations_controller.rb:2]: why did you remove this lines? I see it in a lot of places, restore this files and remove it from the commit

File: app/models/message.rb [hunk 13]
```diff
@@ -0,0 +1,8 @@
+class Message < ApplicationRecord
+  belongs_to :conversation
+  belongs_to :user
+
+  scope :unread, -> { where(read: false) }
+
+  validates :message, presence: { on: :create, message: "can't be blank" }
```
Comment (juanmferreira93) [inline app/models/message.rb:7]: you could remove the message as Martin said before, Rails do it for you

File: config/routes.rb [hunk 14]
```diff
@@ -2,6 +2,9 @@
   devise_for :admin_users, ActiveAdmin::Devise.config
   ActiveAdmin.routes(self)
   namespace :v1 do
+    post '/messages/:id', to: 'messages#create', as: :messages
+    get '/messages/:id', to: 'messages#index'
```
Comment (juanmferreira93) [inline config/routes.rb:6]: why not to use `resources` for this?

File: db/migrate/20210518193548_create_message.rb [hunk 15]
```diff
@@ -0,0 +1,12 @@
+class CreateMessage < ActiveRecord::Migration[6.1]
+  def change
+    create_table :messages do |t|
+      t.integer :conversation_id
+      t.integer :user_id
```
Comment (juanmferreira93) [inline db/migrate/20210518193548_create_message.rb:5]: why you don't create an `association` with `users` here?

File: db/migrate/20210518183834_create_conversations_table.rb [hunk 16]
```diff
@@ -0,0 +1,16 @@
+class CreateConversationsTable < ActiveRecord::Migration[6.1]
+  def change
+    create_table :conversations do |t|
+      t.integer :user_id1
+      t.integer :user_id2
```
Comment (juanmferreira93) [inline db/migrate/20210518183834_create_conversations_table.rb:5]: same here, try to change it to associations

File: spec/factories/conversations_factories.rb [hunk 17]
```diff
@@ -0,0 +1,7 @@
+FactoryBot.define do
+  factory :conversation do
+    user_id1 { create(:user).id }
+    user_id2 { create(:user).id }
```
Comment (juanmferreira93) [inline spec/factories/conversations_factories.rb:4]: once you change the migrations, you probably should change this lines too

File: spec/requests/conversations/conversations_show_spec.rb [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+      it 'returns the conversation' do
+        get v1_conversations_path(logged_user_conversation.id)
+        expect(json['conversations'][0]['id']).to eq(logged_user_conversation.id)
+        expect(json['conversations'][0]['user_id1']).to eq(user1.id)
+      end
+
+      # it 'updates read attr to true' do
+      #   get v1_conversations_path(logged_user_conversation.id)
+      #   expect(logged_user_conversation.reload.read).to eq(true)
+      # end
```
Comment (juanmferreira93) [inline spec/requests/conversations/conversations_show_spec.rb:20]: 👀 , to skip a test you could use `xit` or `xcontext` to skip all the context

File: spec/requests/conversations/conversations_show_spec.rb [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+    end
+
+    context 'when trying to get a conversation the user doesn\'t participate in' do
+      # it 'returns error' do
+      #   get v1_conversations_path(other_conversation.id)
+      #   byebug
+      #   expect(json['error']).to eq()
+      # end
```
Comment (juanmferreira93) [inline spec/requests/conversations/conversations_show_spec.rb:28]: 👀

File: spec/requests/targets/targets_create_spec.rb [hunk 20]
```diff
@@ -72,15 +72,4 @@
...
-  context 'when user is not logged in' do
-    it 'returns http unauthorized' do
-      subject
-      expect(response).to have_http_status(401)
-    end
-
-    it 'doesn\t change the target count' do
-      expect { subject }.not_to change(Target, :count)
-    end
-  end
```
Comment (juanmferreira93) [inline spec/requests/targets/targets_create_spec.rb:85]: why did you remove this specs?

File: app/models/conversation.rb (hunk 1 above)
Comment (ivanfelfo) [reply app/models/conversation.rb:6]: 👍

File: spec/models/conversation_spec.rb (hunk 4 above)
Comment (ivanfelfo) [reply spec/models/conversation_spec.rb:11]: 👍

File: spec/models/conversation_spec.rb (hunk 5 above)
Comment (ivanfelfo) [reply spec/models/conversation_spec.rb:15]: 👍

File: app/controllers/v1/conversations_controller.rb (hunk 6 above)
Comment (ivanfelfo) [reply app/controllers/v1/conversations_controller.rb:11]: 👍

File: app/controllers/v1/conversations_controller.rb (hunk 7 above)
Comment (ivanfelfo) [reply app/controllers/v1/conversations_controller.rb:23]: 👍

File: app/controllers/v1/conversations_controller.rb (hunk 8 above)
Comment (ivanfelfo) [reply app/controllers/v1/conversations_controller.rb:31]: 👍

File: app/controllers/v1/conversations_controller.rb (hunk 9 above)
Comment (ivanfelfo) [reply app/controllers/v1/conversations_controller.rb:43]: 👍

File: app/controllers/v1/messages_controller.rb (hunk 11 above)
Comment (ivanfelfo) [reply app/controllers/v1/messages_controller.rb:27]: 👍

File: app/models/message.rb (hunk 13 above)
Comment (ivanfelfo) [reply app/models/message.rb:7]: 👍

File: spec/requests/targets/targets_create_spec.rb (hunk 20 above)
Comment (ivanfelfo) [reply spec/requests/targets/targets_create_spec.rb:85]: they were repeated 😅

File: db/migrate/20210518183834_create_conversations_table.rb [hunk 21]
```diff
@@ -0,0 +1,16 @@
+class CreateConversationsTable < ActiveRecord::Migration[6.1]
+  def change
+    create_table :conversations do |t|
+      t.integer :user_id1
```
Comment (martinjaimem) [inline db/migrate/20210518183834_create_conversations_table.rb:4]: maybe we should add `null: false` what do you think?

File: db/migrate/20210518193548_create_message.rb [hunk 22]
```diff
@@ -0,0 +1,12 @@
+class CreateMessage < ActiveRecord::Migration[6.1]
+  def change
+    create_table :messages do |t|
+      t.integer :conversation_id
```
Comment (martinjaimem) [inline db/migrate/20210518193548_create_message.rb:4]: What about adding null constraints?

File: db/migrate/20210518183834_create_conversations_table.rb (hunk 21 above)
Comment (ivanfelfo) [reply db/migrate/20210518183834_create_conversations_table.rb:4]: 👍

File: db/migrate/20210518193548_create_message.rb (hunk 22 above)
Comment (ivanfelfo) [reply db/migrate/20210518193548_create_message.rb:4]: 👍

File: db/migrate/20210518193548_create_message.rb (hunk 15 above)
Comment (ivanfelfo) [reply db/migrate/20210518193548_create_message.rb:5]: 👍

File: app/views/v1/conversations/show.json.jbuilder (hunk 3 above)
Comment (ivanfelfo) [reply app/views/v1/conversations/show.json.jbuilder:2]: 👍

File: app/views/v1/messages/index.json.jbuilder [hunk 23]
```diff
@@ -0,0 +1,8 @@
+json.messages do
+  json.array! @messages
```
[the remark this answers is not in the document]
Comment (ivanfelfo) [reply app/views/v1/messages/index.json.jbuilder:2]: 👍

File: app/controllers/v1/conversations_controller.rb [hunk 24]
```diff
@@ -14,12 +14,14 @@ def index
...
 
     def show
-      @conversation = conversation_show_array
-      @conversation.update!(read: true)
-      return unless @conversation.empty?
-
-      render json: { error: I18n.t('conversation.show.empty') }, status: :bad_request
-      @conversation.messages.unread.each { |m| m.read = true } if @conversation.messages.present?
+      @conversation = conversation_show_query
+      # byebug
```
Comment (juanmferreira93) [inline app/controllers/v1/conversations_controller.rb:18]: remove this comment

File: spec/models/conversation_spec.rb [hunk 25]
```diff
@@ -0,0 +1,21 @@
...
+    end
+  end
+
+  context "when it\'s NOT a valid instance" do
+    context 'when trying to create a conversation with the same person' do
+      let(:new_conversation) { build(:conversation, user_id1: 1, user_id2: 1) }
+
+      it 'raises an error' do
+        expect { subject }.to raise_error(ActiveRecord::RecordInvalid)
+        # "Validation failed: User id1 can'\t create a conversation with yourself")
```
Comment (juanmferreira93) [inline spec/models/conversation_spec.rb:17]: remove this comment

**Review by juanmferreira93 — APPROVED** [review]

File: app/controllers/v1/conversations_controller.rb (hunk 24 above)
Comment (ivanfelfo) [reply app/controllers/v1/conversations_controller.rb:18]: 👍

**Review by martinjaimem — APPROVED** [review]
I'll approve this for the sake of deploying the app, however, you should take into account for the future that:
 - no PR should be more than 20+ files
 - no PR should include more than one functionality
